### PR TITLE
APEX-54 #resolve Added code to copy from local file to HDFS with over…

### PR DESCRIPTION
**Pull request by ishark**
…write option for AsyncFSStorageAgent

### CODE REVIEW

File: common/src/main/java/com/datatorrent/common/util/AsyncFSStorageAgent.java [hunk 1]
```diff
@@ -16,18 +16,26 @@
 package com.datatorrent.common.util;
 
 import java.io.File;
+import java.io.FileInputStream;
```
Comment (243826) [inline common/src/main/java/com/datatorrent/common/util/AsyncFSStorageAgent.java]: import java.io.*

File: common/src/main/java/com/datatorrent/common/util/AsyncFSStorageAgent.java [hunk 2]
```diff
@@ -16,18 +16,26 @@
...
+import java.io.InputStream;
 import java.io.ObjectStreamException;
+import java.io.OutputStream;
+import java.util.EnumSet;
 
 import org.apache.hadoop.conf.Configuration;
+import org.apache.hadoop.fs.CreateFlag;
+import org.apache.hadoop.fs.FSDataOutputStream;
 import org.apache.hadoop.fs.FileSystem;
 import org.apache.hadoop.fs.FileUtil;
```
Comment (243826) [inline common/src/main/java/com/datatorrent/common/util/AsyncFSStorageAgent.java]: import fs.* (5 or more)

File: common/src/main/java/com/datatorrent/common/util/AsyncFSStorageAgent.java [hunk 3]
```diff
@@ -86,8 +94,50 @@ public void copyToHDFS(final int operatorId, final long windowId) throws IOExcep
...
+          fileContext.rename(lPath, new Path(path + Path.SEPARATOR + operatorIdStr + Path.SEPARATOR + window), Options.Rename.OVERWRITE);
+        }
+      }
+    }
+  }
+
+  public boolean copyStream(File src, OutputStream out, Configuration conf) throws IOException
+  {
+    if (src.isFile()) {
+      InputStream in = null;
```
Comment (243826) [inline common/src/main/java/com/datatorrent/common/util/AsyncFSStorageAgent.java]: InputStream in = new FileInputStream(src);
try {
}
finally {
in.close();
}

File: common/src/main/java/com/datatorrent/common/util/AsyncFSStorageAgent.java [hunk 4]
```diff
@@ -86,8 +94,50 @@ public void copyToHDFS(final int operatorId, final long windowId) throws IOExcep
...
+      try {
+        in = new FileInputStream(src);
+        IOUtils.copyBytes(in, out, conf);
+      } catch (IOException e) {
+        IOUtils.closeStream(out);
```
Comment (243826) [inline common/src/main/java/com/datatorrent/common/util/AsyncFSStorageAgent.java]: outputStream should be closed by whoever opened it.

File: common/src/main/java/com/datatorrent/common/util/AsyncFSStorageAgent.java [hunk 5]
```diff
@@ -86,8 +94,50 @@ public void copyToHDFS(final int operatorId, final long windowId) throws IOExcep
...
+        IOUtils.closeStream(in);
+        throw e;
+      }
+    } else {
+      throw new IOException(src.toString() + ": No such file");
+    }
+    return FileUtil.fullyDelete(src);
```
Comment (243826) [inline common/src/main/java/com/datatorrent/common/util/AsyncFSStorageAgent.java]: if you are going to delete, probably better to call the method moveFile...

**Comment by tweise** [comment]
Please squash.

File: common/src/main/java/com/datatorrent/common/util/AsyncFSStorageAgent.java [hunk 6]
```diff
@@ -86,8 +83,44 @@ public void copyToHDFS(final int operatorId, final long windowId) throws IOExcep
...
+
+  public boolean moveFile(File src, OutputStream out, Configuration conf) throws IOException
+  {
+    InputStream in = null;
+    try {
+      in = new FileInputStream(src);
+      IOUtils.copyBytes(in, out, conf, false);
+    } finally {
+      IOUtils.closeStream(in);
+    }
```
Comment (gauravgopi123) [inline common/src/main/java/com/datatorrent/common/util/AsyncFSStorageAgent.java]: This needs to be discussed more before merging

Comment (tweise) [inline common/src/main/java/com/datatorrent/common/util/AsyncFSStorageAgent.java]: Gaurav, what is the concern?

**Pull request merged**